keps/0007-dynamic-audit-configuration.md

```diff
@@ -0,0 +1,231 @@
+
+## Summary
+
+We want to allow the advanced auditing features to be dynamically configured. Following in the same vein as [Dynamic Admission Control](https://kubernetes.io/docs/admin/extensible-admission-controllers/) we would like to provide a means of configuring the auditing features post cluster provisioning.
+
+## Motivation
+
+The advanced auditing features are a powerful tool, yet difficult to configure. The configuration requires deep insight into the deployment mechanism of choice and often takes many iterations to configure properly requiring a restart of the apiserver each time. Moreover, the ability to install addon tools that configure and enhance audting is hindered by the overhead in configuration. Such tools frequently run on the cluster requiring future knowledge of how to reach them when the cluster is live. These tools could enhance the security and conformance of the cluster and its applications.
+
+### Goals
+- Provide an api and set of objects to configure the advanced auditing kube-apiserver configuration dynamically
+
+### Non-Goals
+- Provide a generic interface to configure all kube-apiserver flags
+
+## Proposal
+
+### Dynamic Configuration
+Dynamic configuration of the advanced auditing features will be provided through use of a shared informer.
+
+#### Cluster Scoped Configuration
+A cluster scoped configuration object will be provided that applies to all events in the cluster.
+
+```golang
+// ClusterAuditConfiguration represents a cluster level audit configuration
+type ClusterAuditConfiguration struct {
+    metav1.TypeMeta
+
+    v1.ObjectMeta
+
+    // Backends to send events
+    Backends *Backends `json:"backends"`
+}
+
+// Backends holds the configuration for the various event backends
+type Backends struct {
+    // Webhooks holds the webhook backends
+    Webhooks []*WebhookBackend `json:"webhooks,omitempty"`
+}
+
+type WebhookMode string
+
+const (
+    // WebhookModeBatch sends requests in batches
+    // bloking mode is currently disable but will be discussed at a later date
+    WebhookModeBatch WebhookMode = "batch"
+)
+
+// WebhookBackend holds the configuration of the webhooks
+type WebhookBackend struct {
+    // Mode for webhook backend
+    Mode WebhookMode `json:"mode"`
+
+    // InitialBackoff is amount of time to wait before retrying the first failed request in seconds
+    InitialBackoff *int `json:"initialBackoff,omitempty"`
+
+    // BatchConfig for the batch webhook mode
+    BatchConfig *BatchConfig `json:"batch,omitempty"`
+
+    // ClientConfig holds the connection parameters for the webhook
+    ClientConfig *ClientConfig `json:clientConfig"`
+
+    // TruncateOptions follows the truncate options 
+    // https://github.com/kubernetes/apiserver/blob/master/pkg/server/options/audit.go#L157
+    TruncateOptions *AuditTruncateOptions `json:truncateOptions,omitempty"`
+}
+
+// BatchConfig holds the configuration for batch webhooks
+type BatchConfig struct {
```

Comment:
I'm thinking we should aim to keep parity with the existing options if we choose to expose them, otherwise it seems like an odd mismatch between the flags and configuration object. Is there a reason why we want to limit them?

Reply:
Because a lot of those have to do with performance tuning of the apiserver process, which isn't something that users setting up external webhooks should be concerned with, IMO. A few of these do affect the receiving server though, so I could see keeping those (ThrottleBurst, ThrottleQPS, TruncateOptions)

Reply:
ok I could see `BufferSize` and `MaxWait` as only api-server specific. `MaxSize` could potentially be backend specific, unless we find that to be too dangerous. The difficulty with this is that now those flags need to be denoted as global options, are we comfortable changing that here?